### PR TITLE
[clang] pointer to member with qualified-id enclosed in parentheses in unevaluated context should be invalid

### LINKED ISSUE #40906: decltype(&(C::member)) yields pointer-to-member type
|  |  |
| --- | --- |
| Bugzilla Link | [41561](https://llvm.org/bz41561) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @hfinkel,@randomnetcat,@darkbuck,@riccibruno,@zygoloid |

## Extended Description 
In some contexts, decltype(&(C::member)) yields a pointer to member type, but according to [expr.unary.op]/4, no pointer to member should be formed when the operand of the & operator is enclosed in parentheses.

http://eel.is/c++draft/expr.unary.op#4

The code below demonstrates this behavior. It compiles on all Clang versions I have tried between 3.2 and HEAD, even though I don't think it should.

```
#include <type_traits>
 
struct S {
    int x;

    static_assert(std::is_same<decltype(&(S::x)), int S::*>::value, ""); // !

    void foo() {
        static_assert(std::is_same<decltype(&(S::x)), int*>::value, ""); // OK
    }
};

static_assert(std::is_same<decltype(&(S::x)), int S::*>::value, ""); // !

int main() {
}
```

### PR DESCRIPTION
**Pull request by zwuis**
clang don't check whether the operand of the & operator is enclosed in parantheses when pointer to member is formed in unevaluated context, for example:

```cpp
struct foo { int val; };

int main() { decltype(&(foo::val)) ptr; }
```

`decltype(&(foo::val))` should be invalid, but clang accepts it. This PR fixes this issue.

Fixes #40906.

### CODE REVIEW

**Comment by cor3ntin** [comment]
I should note that while i think this is correct per the standard, no one seems to be conforming here

File: clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp [hunk 1]
```diff
@@ -27,3 +26,20 @@ namespace rdar10544564 {
...
+  };
+
+  void test() {
+    decltype(&(A::val)) ptr1; // expected-error {{invalid use of non-static data member 'val'}}
+    int A::* ptr2 = &(A::val); // expected-error {{invalid use of non-static data member 'val'}}
+
+    // FIXME: Error messages in these cases are less than clear, we can do
+    // better.
+    int size = sizeof(&(A::func)); // expected-error {{call to non-static member function without an object argument}}
+    void (A::* ptr3)() = &(A::func); // expected-error {{call to non-static member function without an object argument}}
```
Comment (cor3ntin) [inline clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp:43]: I would expect `err_invalid_non_static_member_use` to be used there too (or rather, that case should be consistent with above)

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -14644,6 +14644,17 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
...
           }
 
+          // C++11 [expr.unary.op] p4:
+          // A pointer to member is only formed when an explicit & is used and
+          // its operand is a qualified-id not enclosed in parentheses.
+          if (isa<ParenExpr>(OrigOp.get())) {
+            // `op->getEndLoc()` is the last part of the qualified-id.
+            // For example, "baz" in "foo::bar::baz".
+            Diag(op->getEndLoc(), diag::err_invalid_non_static_member_use)
+                << dcl->getDeclName() << op->getSourceRange();
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:14654]: Maybe we could have a more specific error, like "cannot form a pointer to member from a parenthesized expression" ?
I would expect the error to point to the opening paren (`OrigOp->getBeginLoc()` presumably)

File: clang/lib/Sema/SemaExpr.cpp [hunk 3]
```diff
@@ -14644,6 +14644,17 @@ QualType Sema::CheckAddressOfOperand(ExprResult &OrigOp, SourceLocation OpLoc) {
...
+            return QualType();
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:14655]: Continuing might lead to better error recovery here

File: clang/lib/Sema/SemaExpr.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaExpr.cpp:14654]: We should also add a fix-it to remove the parentheses as a kindness to the user.

**Review by AaronBallman — COMMENTED** [review]
Thank you for the fix! You should also add a release note to clang/docs/ReleaseNotes.rst so that users know about the improvement.

File: clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp (hunk 1 above)
Comment (zwuis) [reply clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp]: > I would expect `err_invalid_non_static_member_use` to be used there too (or rather, that case should be consistent with above)

This inconsistency is irrelevant to PR 40906. I am confused about the following questions:

1. Do I need to remove these tests?
2. Do I need to fix this inconsistency in this PR, or a new PR?
3. Should I open an issue for it?

**Comment by cor3ntin** [comment]
@t3nsor Do you know the reason for the "not enclosed in parameters" in https://eel.is/c++draft/expr.unary.op#4 ?

**Comment by t3nsor** [comment]
> @t3nsor Do you know the reason for the "not enclosed in parameters" in https://eel.is/c++draft/expr.unary.op#4 ?

No I don't. If I had to guess, it might be because adding parentheses is intended to be used as a way to not create a pointer to member (and instead get an ordinary pointer to a member named by a qualified  name) sort of like how parentheses can be used to suppress ADL.

File: clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp (hunk 1 above)
Comment (zwuis) [reply clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp]: > This inconsistency is irrelevant to issue 40906. I am confused about the following questions:
> 
> 1. Do I need to remove these tests?
> 2. Do I need to fix this inconsistency in this PR, or a new PR?
> 3. Should I open an issue for it?

@cor3ntin Could you give me some ideas for these questions? Thank you!

Comment (cor3ntin) [reply clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp]: I think creating an issue, or a new PR is the way to go. Would that work for you?

**Review by cor3ntin — APPROVED** [review]
Sorry for the delayed reply.
LGTM, thank you for your contribution!

File: clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp (hunk 1 above)
Comment (zwuis) [reply clang/test/CXX/expr/expr.unary/expr.unary.op/p3.cpp]: > I think creating an issue, or a new PR is the way to go. Would that work for you?

Yes. Issue made: #90822. I will create PR in a few days.

**Comment by cor3ntin** [comment]
@zwuis Thanks for your contribution. Will you need me to merge that for you?

**Comment by zwuis** [comment]
> @zwuis Thanks for your contribution. Will you need me to merge that for you?

Yes. Thanks for your review!

**Comment by hubert-reinterpretcast** [comment]
@cor3ntin @AaronBallman, the new diagnostic triggers for valid code (https://gcc.godbolt.org/z/fMvEaGabG):
```cpp
#include <type_traits>
 
struct S {
    int x;
    static_assert(std::is_same<decltype((S::x)), int &>::value, "");
    static_assert(std::is_same<decltype(&(S::x)), int *>::value, "");
};
```

Either the first `static_assert` is invalid or both `static_assert`s are valid.

**Comment by hubert-reinterpretcast** [comment]
Indeed, the new diagnostic is never valid because the permission to name non-static data members in unevaluated operands is not restricted by scope: https://eel.is/c++draft/expr.prim.id.general#4.3

https://gcc.godbolt.org/z/h4eTT984M:
```cpp
namespace std {
  template <typename, typename>
  struct is_same { static constexpr bool value = false; };

  template <typename T>
  struct is_same<T, T> { static constexpr bool value = true; };
}
 
struct S {
    int x;
    static_assert(std::is_same<decltype((S::x)), int &>::value, "");
    static_assert(std::is_same<decltype(&(S::x)), int *>::value, "");
};
static_assert(std::is_same<decltype((S::x)), int &>::value, "");
static_assert(std::is_same<decltype(&(S::x)), int *>::value, "");
```

**Comment by AaronBallman** [comment]
If I'm following along correctly, both of those `static_assert`s should be accepted from your first example because `decltype` is an unevaluated context and the fix would be to change:
```
if (isa<ParenExpr>(OrigOp.get())) {
```
to instead be:
```
if (isa<ParenExpr>(OrigOp.get()) && !SemaRef.isUnevaluatedContext()) {
```
? Or is there a deeper issue here and I'm not seeing it?

**Comment by hubert-reinterpretcast** [comment]
> Or is there a deeper issue here and I'm not seeing it?

The previous Clang behaviour produced the wrong type as the result. So the choices are to:

1. Do the old wrong thing: silent incorrect behaviour
2. Do the new wrong thing: produce diagnostic that may confuse users
3. Fix the type associated with the expression

**Comment by hubert-reinterpretcast** [comment]
The optics of having something in the release notes about this (if the bug is not fixed properly) is not good.
Meaning that I am leaning towards options (1) or (3).

**Comment by AaronBallman** [comment]
Thank you for spotting this, btw!

Given the proximity to the branch date, I lean towards (1) to give us time to figure out how to properly do (3) unless someone is starting on (3) Real Soon Now™. Any disagreement @cor3ntin ?

**Comment by cor3ntin** [comment]
I can try to look at that this week. Either way be ill-formed is much better than being incorrect

**Comment by hubert-reinterpretcast** [comment]
I think (at least part of) the fix is to add a condition to

https://github.com/llvm/llvm-project/blob/21e6777957457451196084cd48ebc42bce9619f0/clang/lib/Sema/SemaExpr.cpp#L14117-L14120

that the operand is not parenthesized. For reference, the address-of-member-function case checks for parentheses this way:

https://github.com/llvm/llvm-project/blob/21e6777957457451196084cd48ebc42bce9619f0/clang/lib/Sema/SemaExpr.cpp#L13908-L13912

This should address the pre-existing bug with:
```cpp
namespace std {
  template <typename, typename>
  struct is_same { static constexpr bool value = false; };

  template <typename T>
  struct is_same<T, T> { static constexpr bool value = true; };
}
 
struct S {
    int &x;
    static_assert(std::is_same<decltype(&(S::x)), int *>::value, "");
};
```

**Comment by zwuis** [comment]
Note that this diagnostic doesn't be emitted in member function.

```cpp
#include <type_traits>

struct S {
    int x;
    void func() {
        // no diagnostic, expected
        static_assert(std::is_same_v<decltype(&(s::x)), int*>, "");
    }
    // diagnostic emitted, unexpected
    static_assert(std::is_same_v<decltype(&(s::x)), int*>, "");
};
```

**Comment by hubert-reinterpretcast** [comment]
For info, the cases where the _id-expression_ is transformed into a class member access expression (`q` and `h` below) do not run into this code (see https://cplusplus.github.io/CWG/issues/2902.html regarding said transformation):
```cpp
#include <type_traits>
struct S {
  int x;

  bool q = std::is_same_v<decltype(&(S::x)), int *>; // OK

  static void f() {
    static_assert(std::is_same_v<decltype(&(S::x)), int *>, ""); // OK
  }
  void g(this const S &) {
    static_assert(std::is_same_v<decltype(&(S::x)), int *>, ""); // OK
  }
  void h() const {
    static_assert(std::is_same_v<decltype(&(S::x)), const int *>, ""); // OK
  }
};

static_assert(S{0}.q);
```

**Comment by AaronBallman** [comment]
> I can try to look at that this week. Either way be ill-formed is much better than being incorrect

Thank you @cor3ntin! If it looks like you won't have a solution by Monday sometime, let's revert the changes so they don't end up on the release branch. We can cherry-pick if we think we have a rock solid solution, if needs be.

**Comment by zwuis** [comment]
These changes pass all tests above:

```diff
-if (isa<DeclRefExpr>(op) && cast<DeclRefExpr>(op)->getQualifier()) {
+if (isa<DeclRefExpr>(op) && cast<DeclRefExpr>(op)->getQualifier() &&
+    !isa<ParenExpr>(OrigOp.get())) {
   DeclContext *Ctx = dcl->getDeclContext();
   if (Ctx && Ctx->isRecord()) {
     if (dcl->getType()->isReferenceType()) {
       Diag(OpLoc,
             diag::err_cannot_form_pointer_to_member_of_reference_type)
         << dcl->getDeclName() << dcl->getType();
       return QualType();
     }
-
-    // C++11 [expr.unary.op] p4:
-    // A pointer to member is only formed when an explicit & is used and
-    // its operand is a qualified-id not enclosed in parentheses.
-    if (isa<ParenExpr>(OrigOp.get())) {
-      SourceLocation LeftParenLoc = OrigOp.get()->getBeginLoc(),
-                      RightParenLoc = OrigOp.get()->getEndLoc();
-
-      Diag(LeftParenLoc,
-            diag::err_form_ptr_to_member_from_parenthesized_expr)
-          << SourceRange(OpLoc, RightParenLoc)
-          << FixItHint::CreateRemoval(LeftParenLoc)
-          << FixItHint::CreateRemoval(RightParenLoc);
-
-      // Continuing might lead to better error recovery.
-    }
 
     while (cast<RecordDecl>(Ctx)->isAnonymousStructOrUnion())
       Ctx = Ctx->getParent();
 
     QualType MPTy = Context.getMemberPointerType(
         op->getType(),
         Context.getTypeDeclType(cast<RecordDecl>(Ctx)).getTypePtr());
     // Under the MS ABI, lock down the inheritance model now.
     if (Context.getTargetInfo().getCXXABI().isMicrosoft())
       (void)isCompleteType(OpLoc, MPTy);
     return MPTy;
   }
 }
```

PS: Sorry for misunderstanding the original issue.

**Comment by hubert-reinterpretcast** [comment]
> PS: Sorry for misunderstanding the original issue.

Totally understandable. Thanks for following up on this.

> These changes pass all tests above:

Can you check this case with your build?
```cpp
struct A { int x; };

char q(int *);
short q(int A::*);

template <typename T>
constexpr int f(char (*)[sizeof(q(&T::x))]) { return 1; }

template <typename T>
constexpr int f(char (*)[sizeof(q(&(T::x)))]) { return 2; }

constexpr int g(char (*p)[sizeof(char)] = 0) { return f<A>(p); }
constexpr int h(char (*p)[sizeof(short)] = 0) { return f<A>(p); }

static_assert(g() == 2);
static_assert(h() == 1);
```

**Comment by zwuis** [comment]
> Can you check this case with your build?

Also pass this test.

**Pull request merged**